File: loop_files/affine_transformations.m
% Code to illustrate the application of affine transformations to 2D polygons
clc
clear all
close all

%   Define a 3*5 matrix made up of five 3*1 column vectors 
%   These vectors are the vertices of the polygon. 

polygon_point(1,1) = 0.0 ;  
polygon_point(2,1) = 0.0 ; 
polygon_point(3,1) = 1.0 ; 
polygon_point(1,2) = 10.0  ; 
polygon_point(2,2) = 0.0 ;  
polygon_point(3,2) = 1.0 ;
polygon_point(1,3) = 10.0 ; 
polygon_point(2,3) = 10.0 ; 
polygon_point(3,3) = 1.0 ;
polygon_point(1,4) = 0.0;  
polygon_point(2,4) = 10.0 ; 
polygon_point(3,4) = 1.0 ;
polygon_point(1,5) = polygon_point(1,1) ;  
polygon_point(2,5) = polygon_point(2,1) ;  
polygon_point(3,5) = 1.0 ;

centre_x = 0.25*(polygon_point(1,1) + polygon_point(1,2) + polygon_point(1,3) + polygon_point(1,4)) ;
centre_y = 0.25*(polygon_point(2,1) + polygon_point(2,2) + polygon_point(2,3) + polygon_point(2,4)) ;

theta = 3*pi/4.0  ; 

rot_matrix  = [ cos(theta)  -sin(theta) 0 ; sin(theta) cos(theta)  0 ; 0 0 1 ]  ;
rot_polygon  = rot_matrix*polygon_point ; 

tx = 2.0 ; 
ty = 4.0 ; 

trans_matrix  = [ 1  0 tx  ;  0 1 ty ;  0 0 1  ]  ;
trans_polygon  = trans_matrix*polygon_point ; 

ref_matrix  = [ -1  0 0  ;  0 1 0 ;  0 0 1  ]  ;
ref_polygon  = ref_matrix*polygon_point ; 

scaling_matrix  = [ 0.5  0 0  ;  0 0.5 0 ;  0 0 1  ]  ;
scaled_polygon  = scaling_matrix*polygon_point ; 

shear_matrix  = [ 1  0.5 0  ;  0 1 0 ;  0 0 1  ]  ;
shear_polygon  = shear_matrix*polygon_point ; 

scaled_then_translated_polygon = trans_matrix*scaling_matrix*polygon_point ; 

figure
plot(polygon_point(1,1:5),polygon_point(2,1:5),'-*') 
grid on
hold on
fill(rot_polygon(1,1:5),rot_polygon(2,1:5),'g-*')  
plot(trans_polygon(1,1:5),trans_polygon(2,1:5),'y-*')  
plot(ref_polygon(1,1:5),ref_polygon(2,1:5),'k-*')  
plot(scaled_polygon(1,1:5),scaled_polygon(2,1:5),'m-*')  
plot(shear_polygon(1,1:5),shear_polygon(2,1:5),'r-*')  
plot(scaled_then_translated_polygon(1,1:5),scaled_then_translated_polygon(2,1:5),'c-*')  
axis([-30 30 -30 30]) 
legend('Original','Rotated','Translated','Reflected','scaled','shear','scaled then translated') ; 

File: My_Code/affine_transformations.c
#include <stdio.h>
#include <math.h>

#define VERTICES 5
#define ROWS 3

// Function to multiply 3x3 matrix with 3xN matrix
void matrix_multiply(double matrix[ROWS][ROWS], double polygon[ROWS][VERTICES], double result[ROWS][VERTICES]) {
    for (int i = 0; i < ROWS; i++) {
        for (int j = 0; j < VERTICES; j++) {
            result[i][j] = 0;
            for (int k = 0; k < ROWS; k++) {
                result[i][j] += matrix[i][k] * polygon[k][j];
            }
        }
    }
}

// Function to print polygon vertices
void print_polygon(const char* name, double polygon[ROWS][VERTICES]) {
    printf("\n%s:\n", name);
    for (int i = 0; i < 2; i++) {  // Only print x and y coordinates
        printf("%c: ", i == 0 ? 'x' : 'y');
        for (int j = 0; j < VERTICES; j++) {
            printf("%8.2f ", polygon[i][j]);
        }
        printf("\n");
    }
}

int main() {
    // Define original polygon (square)
    double polygon[ROWS][VERTICES] = {
        {0, 10, 10, 0, 0},    // x coordinates
        {0, 0, 10, 10, 0},    // y coordinates
        {1, 1, 1, 1, 1}       // homogeneous coordinate
    };

    // Calculate centroid
    double centre_x = (polygon[0][0] + polygon[0][1] + polygon[0][2] + polygon[0][3]) / 4.0;
    double centre_y = (polygon[1][0] + polygon[1][1] + polygon[1][2] + polygon[1][3]) / 4.0;
    printf("Centroid: (%.2f, %.2f)\n", centre_x, centre_y);

    // Rotation matrix (135 degrees)
    double theta = 3.0 * M_PI / 4.0;
    double rot_matrix[ROWS][ROWS] = {
        {cos(theta), -sin(theta), 0},
        {sin(theta), cos(theta), 0},
        {0, 0, 1}
    };
    double rot_polygon[ROWS][VERTICES];
    matrix_multiply(rot_matrix, polygon, rot_polygon);

    // Translation matrix (tx=2, ty=4)
    double trans_matrix[ROWS][ROWS] = {
        {1, 0, 2},
        {0, 1, 4},
        {0, 0, 1}
    };
    double trans_polygon[ROWS][VERTICES];
    matrix_multiply(trans_matrix, polygon, trans_polygon);

    // Reflection matrix (reflect about y-axis)
    double ref_matrix[ROWS][ROWS] = {
        {-1, 0, 0},
        {0, 1, 0},
        {0, 0, 1}
    };
    double ref_polygon[ROWS][VERTICES];
    matrix_multiply(ref_matrix, polygon, ref_polygon);

    // Scaling matrix (scale by 0.5)
    double scaling_matrix[ROWS][ROWS] = {
        {0.5, 0, 0},
        {0, 0.5, 0},
        {0, 0, 1}
    };
    double scaled_polygon[ROWS][VERTICES];
    matrix_multiply(scaling_matrix, polygon, scaled_polygon);

    // Shear matrix (shear factor 0.5 in x-direction)
    double shear_matrix[ROWS][ROWS] = {
        {1, 0.5, 0},
        {0, 1, 0},
        {0, 0, 1}
    };
    double shear_polygon[ROWS][VERTICES];
    matrix_multiply(shear_matrix, polygon, shear_polygon);

    // Combined: scaling then translation
    double temp[ROWS][ROWS];
    double combined_matrix[ROWS][ROWS];
    
    // Multiply trans_matrix * scaling_matrix
    for (int i = 0; i < ROWS; i++) {
        for (int j = 0; j < ROWS; j++) {
            combined_matrix[i][j] = 0;
            for (int k = 0; k < ROWS; k++) {
                combined_matrix[i][j] += trans_matrix[i][k] * scaling_matrix[k][j];
            }
        }
    }
    
    double scaled_then_translated[ROWS][VERTICES];
    matrix_multiply(combined_matrix, polygon, scaled_then_translated);

    // Print all results
    print_polygon("Original Polygon", polygon);
    print_polygon("Rotated Polygon", rot_polygon);
    print_polygon("Translated Polygon", trans_polygon);
    print_polygon("Reflected Polygon", ref_polygon);
    print_polygon("Scaled Polygon", scaled_polygon);
    print_polygon("Sheared Polygon", shear_polygon);
    print_polygon("Scaled then Translated", scaled_then_translated);

    return 0;
}
